### PR TITLE
Add file memory provider

### PR DESCRIPTION
**Pull request by ldorau**
### Description

Add file memory provider.

### Checklist
- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

**Comment by ldorau** [comment]
I will be back on September 2nd.

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: src/provider/provider_file_memory.c [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+    [_UMF_FILE_RESULT_ERROR_PURGE_LAZY_FAILED] = "lazy purging failed",
+    [_UMF_FILE_RESULT_ERROR_PURGE_FORCE_FAILED] = "force purging failed",
+};
+
+static void os_store_last_native_error(int32_t native_error, int errno_value) {
+    TLS_last_native_error.native_error = native_error;
+    TLS_last_native_error.errno_value = errno_value;
+}
+
+static int file_copy_path(const char *in_path, char out_path[NAME_MAX]) {
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_file_memory.c]: Done.

File: src/libumf.map.in [hunk 2]
```diff
@@ -56,6 +56,7 @@ UMF_1.0 {
         umfProxyPoolOps;
         umfPutIPCHandle;
         umfScalablePoolOps;
+        umfFileMemoryProviderOps;
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/libumf.map.in]: Done. Moved to optional symbols (depends on hwloc).

File: include/umf/providers/provider_file_memory.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    UMF_FILE_RESULT_ERROR_ALLOC_FAILED,        ///< Memory allocation failed
+    UMF_FILE_RESULT_ERROR_ADDRESS_NOT_ALIGNED, ///< Allocated address is not aligned
+    UMF_FILE_RESULT_ERROR_FREE_FAILED,         ///< Memory deallocation failed
+    UMF_FILE_RESULT_ERROR_PURGE_LAZY_FAILED,  ///< Lazy purging failed
+    UMF_FILE_RESULT_ERROR_PURGE_FORCE_FAILED, ///< Force purging failed
+} umf_file_memory_provider_native_error_t;
+
+umf_memory_provider_ops_t *umfFileMemoryProviderOps(void);
+
+/// @brief Create default params for os memory provider
```
[the remark this answers is not in the document]
Comment (ldorau) [reply include/umf/providers/provider_file_memory.h]: Done

File: include/umf/providers/provider_file_memory.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#ifndef UMF_FILE_MEMORY_PROVIDER_H
+#define UMF_FILE_MEMORY_PROVIDER_H
+
+#include "umf/memory_provider.h"
```
[the remark this answers is not in the document]
Comment (ldorau) [reply include/umf/providers/provider_file_memory.h]: Done

**Review by bratpiorka — APPROVED** [review]

File: include/umf/providers/provider_file_memory.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#define UMF_FILE_MEMORY_PROVIDER_H
+
+#include <umf/providers/provider_os_memory.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/// @cond
+#define UMF_FILE_RESULTS_START_FROM 2000
```
Comment (lukaszstolarczuk) [inline include/umf/providers/provider_file_memory.h:18]: devdax is using value of 2000, should we use different one here...?

File: src/provider/provider_file_memory.c [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+
+#include <umf.h>
+#include <umf/memory_provider_ops.h>
+#include <umf/providers/provider_file_memory.h>
+
+#define FILE_INIT_SIZE 4096 /* initial size of a file to be mmapped */
+
+#define TLS_MSG_BUF_LEN 1024
+
+typedef struct os_last_native_error_t {
```
Comment (lukaszstolarczuk) [inline src/provider/provider_file_memory.c:32]: `os_` ? (here in other places around native_error

File: README.md [hunk 7]
```diff
@@ -172,6 +172,20 @@ Additionally, required for tests:
...
+The file memory provider does not support the free operation
+(`umfMemoryProviderFree()` always returns `UMF_RESULT_ERROR_NOT_SUPPORTED`),
+so it should be used with a pool manager that will take over
+the managing of the provided memory - for example the jemalloc pool
+with the `disable_provider_free` parameter set to true.
+
+##### Requirements
+
+1) Linux OS
+
```
Comment (lukaszstolarczuk) [inline README.md:188]: should we mention here the limitation of the file path length (`NAME_MAX`)...?

File: test/provider_file_memory.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+#include <umf/memory_provider.h>
+#include <umf/providers/provider_file_memory.h>
+
+using umf_test::test;
+
+#define FILE_PATH ((char *)"tmp_file")
+#define INVALID_PTR ((void *)0x01)
+
+#define ASSERT_IS_ALIGNED(ptr, alignment)                                      \
```
Comment (lukaszstolarczuk) [inline test/provider_file_memory.cpp:17]: I can see this macro is copy-pasted to all these providers - perhaps place it in some test helper file

File: src/provider/provider_file_memory.c [hunk 9]
```diff
@@ -0,0 +1,618 @@
...
+    umf_result_t ret = UMF_RESULT_SUCCESS;
+    int fd;
+
+    if (strncmp(os_ipc_data->path, file_provider->path, NAME_MAX)) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    fd = os_file_open(os_ipc_data->path);
+    if (fd <= 0) {
+        LOG_PERR("opening a shared memory file (%s) failed", os_ipc_data->path);
```
Comment (lukaszstolarczuk) [inline src/provider/provider_file_memory.c:561]: hmm.. I gues we should have negative tests for IPC, like e.g. here, for improper path

File: src/provider/provider_os_memory_posix.c [hunk 10]
```diff
@@ -106,3 +109,48 @@ void os_strerror(int errnum, char *buf, size_t buflen) {
...
+// open a file
+int os_file_open(const char *path) {
+    if (path == NULL) {
+        LOG_ERR("empty path");
+        return -1;
+    }
+
+    int fd = open(path, O_RDWR);
+    if (fd == -1) {
+        LOG_PERR("cannot open the file: %s", path);
```
Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory_posix.c:122]: pls add negative tests for these

File: src/provider/provider_file_memory.c (hunk 9 above)
Comment (ldorau) [reply src/provider/provider_file_memory.c:561]: Hmm ...

File: test/provider_file_memory.cpp (hunk 8 above)
Comment (ldorau) [reply test/provider_file_memory.cpp]: Done

File: README.md (hunk 7 above)
Comment (ldorau) [reply README.md:189]: Done

File: include/umf/providers/provider_file_memory.h (hunk 5 above)
Comment (ldorau) [reply include/umf/providers/provider_file_memory.h]: Done (`3000`)

File: src/provider/provider_os_memory_posix.c (hunk 10 above)
Comment (ldorau) [reply src/provider/provider_os_memory_posix.c:122]: Hmm ...

File: src/provider/provider_file_memory.c (hunk 9 above)
Comment (ldorau) [reply src/provider/provider_file_memory.c]: Done

File: src/provider/provider_os_memory_posix.c (hunk 10 above)
Comment (ldorau) [reply src/provider/provider_os_memory_posix.c:122]: Done

Comment (lukaszstolarczuk) [reply src/provider/provider_os_memory_posix.c:122]: hmm.. what about empty path negative test?

Comment (ldorau) [reply src/provider/provider_os_memory_posix.c:122]: Added.
Done

File: src/provider/provider_file_memory.c [hunk 11]
```diff
@@ -0,0 +1,620 @@
...
+    assert((ptr & (page_size - 1)) == 0);
+    (void)ptr;       // unused in Release build
+    (void)page_size; // unused in Release build
+}
+
+static int file_mmap_aligned(void *hint_addr, size_t length, size_t alignment,
+                             size_t page_size, int prot, int flag, int fd,
+                             size_t *size_fd, os_mutex_t *lock_fd,
+                             void **out_addr, size_t *offset_fd,
+                             size_t *old_offset_fd) {
```
Comment (lplewa) [inline src/provider/provider_file_memory.c:189]: 12 arguments to single function? Can we reduce it?
Also this function is almost ideal copy of this in os memory provider.  (and in #643 it was promissed to be refactored in separate PR)

Refactor ideas:
1. Abstract it to common, as it is needed in multiple places. 
2. Split it to multiple separate functions (i.e. One to grow file, second for mmap)

**Review by lukaszstolarczuk — COMMENTED** [review]
👍

File: src/provider/provider_os_memory_posix.c [hunk 12]
```diff
@@ -106,3 +109,48 @@ void os_strerror(int errnum, char *buf, size_t buflen) {
         LOG_PERR("Retrieving error code description failed");
     }
 }
+
+// open a file
+int os_file_open(const char *path) {
+    if (path == NULL) {
```
Comment (lplewa) [inline src/provider/provider_os_memory_posix.c:115]: nit: !path

File: src/provider/provider_os_memory_posix.c [hunk 13]
```diff
@@ -106,3 +109,48 @@ void os_strerror(int errnum, char *buf, size_t buflen) {
...
+    if (fd == -1) {
+        LOG_PERR("cannot open the file: %s", path);
+    }
+
+    return fd;
+}
+
+// open a file or create
+int os_file_open_or_create(const char *path, size_t size) {
+    if (path == NULL) {
```
Comment (lplewa) [inline src/provider/provider_os_memory_posix.c:130]: nit: !path

File: src/utils/utils_common.h [hunk 14]
```diff
@@ -20,6 +20,8 @@
 extern "C" {
 #endif
 
+#define NAME_MAX 255
```
Comment (lplewa) [inline src/utils/utils_common.h:23]: second definition on name_max. First is in os_provider_internal

File: src/provider/provider_os_memory_posix.c [hunk 15]
```diff
@@ -106,3 +109,48 @@ void os_strerror(int errnum, char *buf, size_t buflen) {
...
+        LOG_ERR("empty path");
+        return -1;
+    }
+
+    int fd = open(path, O_RDWR);
+    if (fd == -1) {
+        LOG_PERR("cannot open the file: %s", path);
```
Comment (lplewa) [inline src/provider/provider_os_memory_posix.c:137]: it's not an error.

File: src/provider/provider_os_memory_posix.c [hunk 16]
```diff
@@ -106,3 +109,48 @@ void os_strerror(int errnum, char *buf, size_t buflen) {
...
+
+        fd = open(path, O_RDWR | O_CREAT | O_EXCL, 0600);
```
Comment (lplewa) [inline src/provider/provider_os_memory_posix.c:139]: Do we two separate opens instead one with O_CREAT (which creates files  if it not exists)? If yes we need comment explaining why.

**Review by lplewa — CHANGES_REQUESTED** [review]

File: src/provider/provider_os_memory_posix.c (hunk 15 above)
Comment (ldorau) [reply src/provider/provider_os_memory_posix.c]: It is now.

File: src/provider/provider_os_memory_posix.c (hunk 13 above)
Comment (ldorau) [reply src/provider/provider_os_memory_posix.c]: Done

File: src/provider/provider_os_memory_posix.c (hunk 16 above)
Comment (lplewa) [reply src/provider/provider_os_memory_posix.c:139]: There is still two opens calls?

Comment (ldorau) [reply src/provider/provider_os_memory_posix.c]: It is `os_file_open_or_create_ORIG`. The correct one is below - I forgot to remove the previous version.
Done now.

Comment (ldorau) [reply src/provider/provider_os_memory_posix.c]: Done

File: src/provider/provider_file_memory.c (hunk 11 above)
Comment (ldorau) [reply src/provider/provider_file_memory.c]: Done

**Comment by ldorau** [comment]
@lplewa all issues fixed, please re-review

File: include/umf/providers/provider_file_memory.h [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+#endif
+
+/// @cond
+#define UMF_FILE_RESULTS_START_FROM 3000
+/// @endcond
+
+/// @brief Memory provider settings struct
+typedef struct umf_file_memory_provider_params_t {
+    /// a path to the file
+    char *path;
```
Comment (lplewa) [inline include/umf/providers/provider_file_memory.h:24]: const

**Review by lplewa — CHANGES_REQUESTED** [review]

File: src/provider/provider_os_memory_posix.c (hunk 16 above)
Comment (ldorau) [reply src/provider/provider_os_memory_posix.c]: Done.

@lplewa ping

File: src/provider/provider_file_memory.c (hunk 11 above)
Comment (ldorau) [reply src/provider/provider_file_memory.c]: Done.

@lplewa ping

File: src/provider/provider_file_memory.c [hunk 18]
```diff
@@ -0,0 +1,656 @@
...
+
+    if (provider == NULL || params == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_file_memory_provider_params_t *in_params =
+        (umf_file_memory_provider_params_t *)params;
+
+    size_t page_size;
+    umf_result_t result = file_get_min_page_size(provider, NULL, &page_size);
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_file_memory.c:97]: Not by accident - it works, because the `provider` argument is not used inside `file_get_min_page_size()`, so I could pass any non-NULL pointer here, but I agree that `os_get_pagesize()` is much better in this case.

Replaced with `os_get_pagesize()`.
Done.

File: src/provider/provider_file_memory.c [hunk 19]
```diff
@@ -0,0 +1,656 @@
...
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    memset(file_provider, 0, sizeof(*file_provider));
+
+    file_provider->page_size = page_size;
+
+    ret = file_translate_params(in_params, file_provider);
+    if (ret != UMF_RESULT_SUCCESS) {
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_file_memory.c:119]: Done

File: include/umf/providers/provider_file_memory.h (hunk 17 above)
Comment (ldorau) [reply include/umf/providers/provider_file_memory.h:24]: Done

**Review by lplewa — CHANGES_REQUESTED** [review]

File: src/provider/provider_os_memory_posix.c [hunk 20]
```diff
@@ -145,3 +145,44 @@ int os_devdax_open(const char *path) {
...
+    int fd = open(path, O_RDWR);
+    if (fd == -1) {
+        LOG_PERR("cannot open the file: %s", path);
+    }
+
+    return fd;
+}
+
+// open a file or create
+int os_file_open_or_create(const char *path, size_t size) {
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_os_memory_posix.c]: Fixed.
Done.

File: src/provider/provider_file_memory.c [hunk 21]
```diff
@@ -0,0 +1,651 @@
...
+        key = rkey;
+    }
+
+    util_mutex_destroy_not_free(&file_provider->lock);
+    critnib_delete(file_provider->fd_offset_map);
+    critnib_delete(file_provider->mmaps);
+    umf_ba_global_free(file_provider);
+}
+
+static inline void assert_is_aligned(uintptr_t ptr, size_t alignment) {
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_file_memory.c]: Done

File: src/provider/provider_file_memory.c [hunk 22]
```diff
@@ -0,0 +1,651 @@
...
+    uintptr_t key = 0;
+    uintptr_t rkey = 0;
+    void *rvalue = NULL;
+    while (1 ==
+           critnib_find(file_provider->mmaps, key, FIND_G, &rkey, &rvalue)) {
+        os_munmap((void *)rkey, (size_t)rvalue);
+        critnib_remove(file_provider->mmaps, rkey);
+        key = rkey;
+    }
+
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_file_memory.c:189]: Done

File: src/provider/provider_file_memory.c [hunk 23]
```diff
@@ -0,0 +1,651 @@
...
+    if (result != UMF_RESULT_SUCCESS) {
+        LOG_ERR("incorrect memory visibility flag: %u", in_params->visibility);
+        return result;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static umf_result_t file_get_min_page_size(void *provider, void *ptr,
+                                           size_t *page_size);
```
[the remark this answers is not in the document]
Comment (ldorau) [reply src/provider/provider_file_memory.c]: Removed.
Done.

File: include/umf/providers/provider_file_memory.h [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+    UMF_FILE_RESULT_ERROR_ALLOC_FAILED,       ///< Memory allocation failed
+    UMF_FILE_RESULT_ERROR_FREE_FAILED,        ///< Memory deallocation failed
+    UMF_FILE_RESULT_ERROR_PURGE_FORCE_FAILED, ///< Force purging failed
+} umf_file_memory_provider_native_error_t;
+
+umf_memory_provider_ops_t *umfFileMemoryProviderOps(void);
+
+/// @brief Create default params for the file memory provider
+static inline umf_file_memory_provider_params_t
+umfFileMemoryProviderParamsDefault(char *path) {
```
[the remark this answers is not in the document]
Comment (ldorau) [reply include/umf/providers/provider_file_memory.h]: Done

File: include/umf/providers/provider_file_memory.h [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+#endif
+
+/// @cond
+#define UMF_FILE_RESULTS_START_FROM 3000
+/// @endcond
+
+/// @brief Memory provider settings struct
+typedef struct umf_file_memory_provider_params_t {
+    /// a path to the file
+    const char *path;
```
[the remark this answers is not in the document]
Comment (ldorau) [reply include/umf/providers/provider_file_memory.h:24]: Done

**Comment by ldorau** [comment]
@lplewa all issues fixed, please re-review

**Comment by ldorau** [comment]
@lplewa all issues fixed, please re-review

**Review by lplewa — APPROVED** [review]
